compiler-tests/src/rust_masm_tests/mod.rs

```diff
@@ -0,0 +1,36 @@
+use expect_test::expect_file;
+use miden_compiler_tests_rust_source::div_u::div_u;
+use miden_hir::Felt;
+use miden_hir::StarkField;
+use proptest::prelude::*;
+use proptest::test_runner::TestRunner;
+
+use crate::execute_emulator;
+use crate::execute_vm;
+use crate::CompTest;
+
+#[test]
+fn u32_div() {
+    // Test expected compilation artifacts
+    let mut test = CompTest::new();
+    test.rust_source("compiler-tests-rust-source", "div_u_app")
+        .expect_wasm(expect_file!["./expected/div_u.wat"])
+        .expect_ir(expect_file!["./expected/div_u.mir"])
```

Comment:
*Question*: Is the expectation that we emit all of these artifacts for every program? I think we'll in general want to assert specific things about the output of each test program, rather than just expecting whatever the compiler emits (as that will be subject to regular change due to optimizations and such). 

For a subset of core fundamental programs I think it's fine to emit everything and be very precise about what each artifact should look like, much like the `lit` tests are designed to do for the IR, but for most programs I would expect us to be testing the output of running the program via the emulator or the VM, rather than caring about the contents of the `.wat`, `.hir`, or `.masm`. At the end of the day, we want to validate that executing the code does what it is supposed to do, and throw enough programs at it that we have confidence that we're not just testing the happy paths.

Speaking of `lit`/`filecheck`, I would like to make sure it is part of the integration test picture. It is primarily designed for testing transformations, but it can also be used to test changes in representation (e.g. `.hir` -> `.masm`). We have the framework for it set up, but its just waiting on `midenc` to be finished up (which I should have done by Monday).

Reply:
> _Question_: Is the expectation that we emit all of these artifacts for every program? I think we'll in general want to assert specific things about the output of each test program, rather than just expecting whatever the compiler emits (as that will be subject to regular change due to optimizations and such).

Oh, no, this is the example of a "check everything" scenario (I should've made more). We can implement all sorts of checks in `CompilerTest`, like `contains`, `match_regex`, and whatnot, even non-text checks using IR nodes/trees. The source is not limited to Rust and could be loaded from Wasm or IR to test only part of the pipeline. In a test, we can pick only those checks that are interested or even none at all. The builder pattern was the wrong choice here; we don't need to shove it all into one line. Here is what it should look like:

```rust
test.rust_source(...);
test.expect_ir(expect_file!["./expected/file.mir"]);
``` 

or 


```rust
test.rust_source(...);
test.expect_ir_contains("...");
test.expect_masm_contains("...");
``` 

or skipping the Rust and starting from Wasm source:


```rust
test.wasm_source(...);
test.expect_ir_contains("...");
``` 

or even starting from IR:


```rust
test.ir_source(...);
test.expect_masm_contains("...");
```

> For a subset of core fundamental programs I think it's fine to emit everything and be very precise about what each artifact should look like, much like the `lit` tests are designed to do for the IR, but for most programs I would expect us to be testing the output of running the program via the emulator or the VM, rather than caring about the contents of the `.wat`, `.hir`, or `.masm`. At the end of the day, we want to validate that executing the code does what it is supposed to do, and throw enough programs at it that we have confidence that we're not just testing the happy paths.

I agree. Most of the programs will not have any checks of the Wasm, IR, or MASM code. I have high hopes on https://github.com/bytecodealliance/wasm-tools/tree/main/crates/wasm-smith to fuzz the hell out of it.

> 
> Speaking of `lit`/`filecheck`, I would like to make sure it is part of the integration test picture. It is primarily designed for testing transformations, but it can also be used to test changes in representation (e.g. `.hir` -> `.masm`). We have the framework for it set up, but its just waiting on `midenc` to be finished up (which I should have done by Monday).

Of course! I'm looking forward to start using it.

Reply:
Got it, that sounds good to me! Definitely interested to see what we can do with `wasm-smith` and `proptest` to generate programs for testing!